Egress QoS support (#5425)
antrea-io/antrea: docs/egress.md

  - [AppliedTo](#appliedto)
  - [EgressIP](#egressip)
  - [ExternalIPPool](#externalippool)

>   - [TrafficShaping](#trafficshaping)

Reviewer: The other headers in this section use the field name, it should do the same here.

Author: Changed.

        role: web
  egressIP: 10.10.0.8 # can be populated by Antrea after assigning an IP from the pool below
  externalIPPool: prod-external-ip-pool

>   rateLimit:

Reviewer 2: IMO we should remove this here and move it to the TrafficShaping section, at least while it is still "Alpha"

Author: Thanks. Done.

    peak: 300

>     burst: 500

Reviewer: I rarely see similar bandwidth configurations have peak and burst (and don't both of them mean burst?), the common configuration is just one value, the maximum bandwidth, like Kubernetes bandwidth annotation `kubernetes.io/egress-bandwidth: 10M`, I wonder if we should just have a field `bandwidth: xx`.
And to make the API more straightforward, could we also support different units, like the annotation, 10K, 10M, 10G?
cc @jianjuns @antoninbas @salv-orlando for opinions.

Reviewer 2: I agree that "peak" is a bit confusion, as opposed to "rate" (or maybe "bandwidth"). I also agree that we should support strings instead of integers, as users will be able to provide the unit and it will be more user-friendly.

However, I feel like we should not follow the model of the K8s annotation, which provides the ability to configure the rate without the ability to configure the burst:
- The K8s bandwidth support has been "experimental" for years, so maybe not a great model to follow
- Setting the burst when configuring the CNI [bandwidth plugin](https://www.cni.dev/plugins/current/meta/bandwidth/) / TC is left to the discretion of the container runtime. Containerd sets it to `MaxUint32` (https://github.com/containerd/containerd/blob/main/pkg/cri/server/sandbox_run.go#L548-L556), which _seems_ to translate to 256Mb in TC
- Some users complain about the lack of control over the burst size: https://github.com/kubernetes/kubernetes/issues/90838. This is very limiting for users who want to do traffic shaping

A configuration like this would be nice IMO:
```yaml
bandwidth: #could be called trafficShaping as well
    rate: 10M # 10Mbps
    burst: 32M #32Mb
```

Reviewer: @antoninbas's API proposal makes sense to me.

Author: Changed.

Reviewer 2: I believe we agreed to use strings with units? The current example is missing units

Reviewer: Could we use the term "bandwidth"? My understand is "traffic shaping" is the method while "bandwidth" is the expectation. For a declarative API, I think it's better to declare the desired state, not how it's achieved. Like how cpu and memory is configured in Pod API, it wouldn't mention "cgroup" or "resource management".
```
  containers:
  - name: app
    resources:
      requests:
        memory: "64Mi"
        cpu: "250m"
```

Reviewer: I only mean the API, using "traffic shaping" in the feature gate name and documentation is fine by me.

Author: Got you. Changed.

be assigned to. It can be empty, which means users should assign the `egressIP`
to one Node manually.

### TrafficShaping

> The `bandwidth` field impose rate-limiting on egress traffic of the Egress. `rate`

Author: Done.

> specifies the maximum transmission rate. `burst` specifies maximum burst for throttle.

Author: Changed.

> `rate` and `burst` should follow the k8s Quantity definition

Author: Done.

[here](https://kubernetes.io/docs/reference/kubernetes-api/common-definitions/quantity/),
e.g. 300k, 100M, 2G. All backend workloads selected by a rate-limited Egress share the

> same bandwidth while sending egress traffic via this Egress.

Reviewer 2: I think we should explicitly state that "if these limits are exceeded, the traffic will be dropped"

Author: Added.

> **Note**: Traffic shaping is current in alpha version. To use this feature, users should

Reviewer: ```suggestion
**Note**: Traffic shaping is currently in alpha version. To use this feature, users should
```

Author: Done.

> enable `enableTrafficShaping` feature gate. Each Egress IP can be applied one bandwidth only.

Reviewer: ```suggestion
enable `EgressTrafficShaping` feature gate. Each Egress IP can be applied one bandwidth only.
```

Author: Done.

> enable `EgressTrafficShaping` feature gate. Each Egress IP can be applied one bandwidth only.

Author: Done.

If multiple Egresses use the same IP but configure different bandwidths, the effective

> bandwidth will be selected randomly. The effective use of the `bandwidth`

Author: Done.

> function hinges on the prerequisite of OVS meter support.

Author: Done.

An Egress with traffic shaping example:

```yaml
apiVersion: crd.antrea.io/v1beta1
kind: Egress
metadata:
  name: egress-prod-web
spec:
  appliedTo:
    namespaceSelector:
      matchLabels:
        env: prod
    podSelector:
      matchLabels:
        role: web

>   egressIP: 10.10.0.8 # can be populated by Antrea after assigning an IP from the pool below

Reviewer: there is no pool below

Author: Done.

be assigned to. It can be empty, which means users should assign the `egressIP`
to one Node manually.

> ### TrafficShaping

Author: Done.

The `bandwidth` field enables traffic shaping for an Egress, by limiting the
bandwidth for all egress traffic belonging to this Egress. `rate` specifies
the maximum transmission rate. `burst` specifies the maximum burst size when
traffic exceeds the rate. The user-provided values for `rate` and `burst` must

> follow the k8s [Quantity](https://kubernetes.io/docs/reference/kubernetes-api/common-definitions/quantity/) format.

Author: Done.

> follow the k8s [Quantity](https://kubernetes.io/docs/reference/kubernetes-api/common-definitions/quantity/) format,

Reviewer 3: `K8s `or `Kubernetes`

Author: Done.

be assigned to. It can be empty, which means users should assign the `egressIP`
to one Node manually.

### RateLimit

> The `rateLimit` field specifies north-south egress traffic of the Egress. `peak`

Reviewer 4: ```suggestion
The `rateLimit` field impose rate-limiting on egress traffic of the Egress. `peak`
```

Author: Done.

specifies the maximum mbps. `burst` specifies maximum burst mbps for throttle.
All backend workloads selected by a rate-limited Egress share the same bandwidth
while sending egress traffic via this Egress.

> **Note**: An Egress with `rateLimit` specified cannot share EgressIP with any

Author: Added a note.

be assigned to. It can be empty, which means users should assign the `egressIP`
to one Node manually.

> ### RateLimit

Reviewer: From the lessons of FQDN, I think we need an alpha feature gate for this support of this feature, otherwise user would misunderstand it shares the maturity of Egress feature.

Author: Added a toggle.

Reviewer 2: A config toggle is different from a feature gate. Also, you are setting the toggle to true by default, which defeats a bit the purpose of "protecting" users from an experimental feature.

I think Quan is suggesting a dedicated feature gate here, and I tend to agree with him. The toggle IMO may not be necessary since the feature is API driven and users should just be able to omit the `rateLimit` field altogether for all their Egress resources if they don't want to use the feature.

At least it feels like we should start with just the feature gate. When we graduate the feature gate, we could consider introducing the toggle, but I am not convinced it will be necessary.

Reviewer: I'm thinking exactly the same as Antonin. No config toggle, just an alpha feature gate, like `EgressTrafficShaping`

Author: Ah I see. Let me change it.

Author: Done.

The `rateLimit` field impose rate-limiting on egress traffic of the Egress. `peak`
specifies the maximum mbps. `burst` specifies maximum burst mbps for throttle.
All backend workloads selected by a rate-limited Egress share the same bandwidth

> while sending egress traffic via this Egress.

Reviewer: Not sure if it's just me, I thought the sentence means every workload has its own exclusive bandwidth and the value is same until I see the code.
